File: OP_lab/lab2/c++/Text.cpp
//
// Created by maxim on 24.02.2020.
//

#include "Text.h"

Text::Text() : quantity{0}, capasity{2} {
    text = new myString[capasity];
    for (int i = 0; i < capasity; ++i)
        text[i] = myString();
}

void Text::print() {
    for (int i = 0; i < quantity; ++i) {
        std::cout << text[i] << std::endl;
    }
}

int Text::get_quantity() {
    return quantity;
}

Text::~Text() {
    delete[] text;
}

void Text::resize(int value) {
    capasity = value + 2;
    auto *tmp = new myString[capasity];
    for (int i = 0; i < capasity; ++i) {
        tmp[i] = i < quantity ? myString(text[i]) : myString();
    }
    if(this->text != nullptr) {
        delete[] text;
    }
    text = tmp;
}

void Text::push_back(char *arr) {
    if(quantity == capasity)
        resize(quantity + 1);
    text[quantity] = myString(arr);
    ++quantity;
}

void Text::push_back(const myString &str) {
    push_back(str.str);
}

void Text::erase(int index) {
    for (int i = index; i < quantity; ++i) {
        text[i] = text[i + 1];
    }
    --quantity;
}

void Text::replace(int index, char *str) {
    text[index].replace(str);
}

void Text::replace(int index, const myString &str) {
    text[index].replace(str);
}

void Text::clear() {
    for (int i = 0; i < quantity; ++i) {
        text[i] = myString();
    }
    quantity = 0;
}

myString Text::return_numbers() {
    myString str;
    for (int i = 0; i < quantity; ++i) {
        str.push_back(text[i].return_numbers());
    }
    return str;
}

myString &Text::at(int index) {
    if(index < 0 || index >= quantity){
        throw "invalid index";
    }
    return text[index];
}

File: OP_lab/lab2/c#/Text.cs
using System;
namespace lab_2_OP_
{
    public class Text
    {
        private MyString[] _text;
        private int _capacity;
        private int _quantity;
        private void Resize(int value)
        {
            _capacity = value + 2;
            MyString[] tmp = new MyString[_capacity];
            for (int i = 0; i < (_text.Length > _capacity ? _text.Length : _capacity); i++)
            {
                if (i < _text.Length)
                {
                    tmp[i] = _text[i];
                }
                else
                {
                    tmp[i] = new MyString();
                }
            }

            _text = tmp;
        }
        
        public Text()
        {
            _capacity = 2;
            _quantity = 0;
            _text = new MyString[_capacity];
            for (int i = 0; i < _capacity; i++)
            {
                _text[i] = new MyString();
            }
        }

        public void Append(string str)
        {
            if (_quantity == _capacity)
            {
                Resize(_text.Length + 1);
            }
            _text[_quantity].push_back(str);
            ++_quantity;
        }
        public void Append(MyString str)
        {
            if (_quantity == _capacity)
            {
                Resize(_text.Length + 1);
            }
            _text[_quantity].push_back(str);
            ++_quantity;
        }
        public void Append(char ch)
        {
            if (_quantity == _capacity)
            {
                Resize(_text.Length + 1);
            }
            _text[_quantity].push_back(ch);
            ++_quantity;
        }

        public void Erase(int index)
        {
            for (int i = index; i < _quantity; i++)
            {
                _text[i] = _text[i + 1];
            }

            --_quantity;
        }

        public void Print()
        {
            for (int i = 0; i < _quantity; i++)
            {
                Console.WriteLine(_text[i]);
            }
        }

        public void Replace(int index, MyString str)
        {
            _text[index].Replace(str);
        }
        public void Replace(int index, string str)
        {
            _text[index].Replace(str);
        }

        public void Clear()
        {
            _quantity = 0;
            for (int i = 0; i < _capacity; i++)
            {
                _text[i] = new MyString();
            }
        }

        public MyString ReturnNumbers()
        {
            MyString tmp = new MyString();
            for (int i = 0; i < _quantity; i++)
            {
                tmp.push_back(_text[i].ReturnNumbers());
            }

            return tmp;
        }

        public int GetQuantity()
        {
            return _quantity;
        }

        public ref MyString at(int index)
        {
            if (index < 0 || index >= _quantity)
            {
                throw new Exception("invalid index");
            }

            return ref _text[index];
        }
    }
}